File: tools/fixtures/rv32-elf-atomic/atomic.c
/*
 * The Compukter Kraft Developers
 *
 * Copyright (C) 2026 Vsevolod Petrov (lazyhat)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

typedef unsigned int u32;

#define CONTROL_STATUS (*(volatile u32 *)0x10000000u)
#define CONTROL_PANIC_CODE (*(volatile u32 *)0x10000004u)
#define CONTROL_EXIT_CODE (*(volatile u32 *)0x10000008u)
#define DEBUG_WRITE (*(volatile unsigned char *)0x10000100u)

static u32 atomic_word = 41u;
static const volatile unsigned char marker[] = "RV32 ELF ATOMIC OK\n";

static __attribute__((noreturn)) void panic(u32 code) {
    CONTROL_PANIC_CODE = code;
    CONTROL_STATUS = 4u;
    for (;;) {
    }
}

__attribute__((noreturn)) void atomic_main(void) {
    u32 old = __atomic_fetch_add(&atomic_word, 1u, __ATOMIC_ACQ_REL);
    if (old != 41u) {
        panic(1u);
    }
    if (__atomic_load_n(&atomic_word, __ATOMIC_ACQUIRE) != 42u) {
        panic(2u);
    }

    __asm__ volatile("fence iorw, iorw" ::: "memory");
    __asm__ volatile("fence.i" ::: "memory");

    for (u32 index = 0; index + 1u < sizeof(marker); ++index) {
        DEBUG_WRITE = (unsigned char)marker[index];
    }
    CONTROL_EXIT_CODE = 0u;
    CONTROL_STATUS = 3u;
    for (;;) {
    }
}

File: host/compukter-vm/src/rv32im/atomic.rs
/*
 * The Compukter Kraft Developers
 *
 * Copyright (C) 2026 Vsevolod Petrov (lazyhat)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct MemoryOrdering {
    pub(crate) acquire: bool,
    pub(crate) release: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum AtomicOp {
    Swap,
    Add,
    Xor,
    And,
    Or,
    Min,
    Max,
    MinU,
    MaxU,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct Rv32Reservation {
    address: u32,
}

impl Rv32Reservation {
    pub(crate) fn new(address: u32) -> Self {
        Self { address }
    }

    pub(crate) fn matches(self, address: u32) -> bool {
        self.address == address
    }

    pub(crate) fn intersects(self, address: u32, size: u32) -> bool {
        let reservation_start = u64::from(self.address);
        let reservation_end = reservation_start + 4;
        let write_start = u64::from(address);
        let write_end = write_start + u64::from(size);
        write_start < reservation_end && reservation_start < write_end
    }
}

pub(crate) fn apply_atomic(operation: AtomicOp, old: u32, operand: u32) -> u32 {
    match operation {
        AtomicOp::Swap => operand,
        AtomicOp::Add => old.wrapping_add(operand),
        AtomicOp::Xor => old ^ operand,
        AtomicOp::And => old & operand,
        AtomicOp::Or => old | operand,
        AtomicOp::Min => {
            if (old as i32) < (operand as i32) {
                old
            } else {
                operand
            }
        }
        AtomicOp::Max => {
            if (old as i32) > (operand as i32) {
                old
            } else {
                operand
            }
        }
        AtomicOp::MinU => old.min(operand),
        AtomicOp::MaxU => old.max(operand),
    }
}
